src/hotspot/share/gc/shenandoah/shenandoahClosures.hpp

```diff
@@ -89,35 +89,30 @@ class ShenandoahUpdateRefsClosure: public ShenandoahOopClosureBase {
   inline void do_oop_work(T* p);
 };
 
-template <DecoratorSet MO = MO_UNORDERED>
-class ShenandoahEvacuateUpdateMetadataClosure: public ShenandoahOopClosureBase {
-private:
-  ShenandoahHeap* const _heap;
-  Thread* const         _thread;
+template <bool atomic>
+class ShenandoahEvacuateUpdateRootClosureBase : public ShenandoahOopClosureBase {
+protected:
+    ShenandoahHeap* const _heap;
 public:
-  inline ShenandoahEvacuateUpdateMetadataClosure();
-  inline void do_oop(oop* p);
-  inline void do_oop(narrowOop* p);
-
-private:
-  template <class T>
-  inline void do_oop_work(T* p);
+    inline ShenandoahEvacuateUpdateRootClosureBase();
+    inline virtual void do_oop(oop* p);
+    inline virtual void do_oop(narrowOop* p);
```

Comment:
Do we have to make these functions `virtual`? I supposed we do templated closures exactly to avoid any sort of virtual method dispatching. Maybe I am missing something here...

Reply:
Fixed

Reply:
Removed `virtual` keyword